private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go

```diff
@@ -0,0 +1,244 @@
+			// We've already loaded this
+			continue
+		}
+		blobPath := normalpath.Join(blobBasedir, hexDigest[:2], hexDigest[2:])
+		contents, err := os.ReadFile(blobPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				return nil, storage.NewErrNotExist(path)
+			}
+			return nil, err
+		}
+		blob, err := manifest.NewMemoryBlob(*digest, contents, manifest.MemoryBlobWithDigestValidation())
+		if err != nil {
+			return nil, err
+		}
+		blobs = append(blobs, blob)
+	}
+	blobSet, err := manifest.NewBlobSet(ctx, blobs)
+	if err != nil {
+		return nil, storage.NewErrNotExist(manifestPath)
+	}
+	bucket, err := manifest.NewBucket(
+		*cacheManifest,
+		*blobSet,
+		manifest.BucketWithAllManifestBlobsValidation(),
+		manifest.BucketWithNoExtraBlobsValidation(),
+	)
+	if err != nil {
+		return nil, storage.NewErrNotExist(manifestPath)
+	}
+	return bufmodule.NewModuleForBucket(ctx, bucket, bufmodule.ModuleWithManifestAndBlobs(*cacheManifest, *blobSet))
+}
+
+func (c *casModuleCacher) PutModule(
+	ctx context.Context,
+	modulePin bufmoduleref.ModulePin,
+	module bufmodule.Module,
+) (retErr error) {
+	moduleManifest := module.Manifest()
+	manifestBlob, err := moduleManifest.Blob()
+	if err != nil {
+		return err
+	}
+	digest := manifestBlob.Digest()
+	if digest == nil {
+		return errors.New("invalid manifest digest")
+	}
+	if modulePinDigestEncoded := modulePin.Digest(); modulePinDigestEncoded != "" {
+		modulePinDigest, err := manifest.NewDigestFromString(modulePinDigestEncoded)
+		if err != nil {
+			return fmt.Errorf("invalid digest %q: %w", modulePinDigestEncoded, err)
+		}
+		if digest.Hex() != modulePinDigest.Hex() {
+			return fmt.Errorf("manifest digest mismatch: expected=%q, found=%q", modulePinDigest.Hex(), digest.Hex())
+		}
+	}
+	moduleBasedir := normalpath.Join(c.baseDir, modulePin.Remote(), modulePin.Owner(), modulePin.Repository())
+	if err := os.MkdirAll(moduleBasedir, 0755); err != nil {
+		return err
+	}
+	if err := c.atomicWrite(strings.NewReader(manifestBlob.Digest().Hex()), normalpath.Join(moduleBasedir, commitsDir, modulePin.Commit())); err != nil {
+		return err
+	}
+	manifestsParentDir := normalpath.Join(moduleBasedir, manifestsDir)
+	if err := c.writeBlob(ctx, manifestBlob, manifestsParentDir); err != nil {
+		return err
+	}
+	blobsParentDir := normalpath.Join(moduleBasedir, blobsDir)
+	writtenDigests := make(map[string]struct{})
+	for _, path := range moduleManifest.Paths() {
+		blobDigest, found := moduleManifest.DigestFor(path)
+		if !found {
+			return fmt.Errorf("failed to find digest for: %q", path)
+		}
+		blobHexDigest := blobDigest.Hex()
+		blobs := module.Blobs()
+		if _, ok := writtenDigests[blobHexDigest]; ok {
+			// We've already written this blob
+			continue
+		}
+		blob, found := blobs.BlobFor(blobDigest.String())
+		if !found {
+			return fmt.Errorf("blob not found for path=%q, digest=%q", path, blobHexDigest)
+		}
+		if err := c.writeBlob(ctx, blob, blobsParentDir); err != nil {
+			return err
+		}
```

Comment:
We could skip this if the blob's already written

Reply:
We don't do this - otherwise the cache couldn't heal itself if one of the files didn't match the digest.

Reply:
Optimized this to first validate if the cache entry matches the digest, and if so we skip the write.